src/tokens/ERC721.sol

```diff
@@ -0,0 +1,312 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity >=0.8.0;
+
+/// @notice Modern and gas efficient ERC-721 + ERC-20/EIP-2612-like implementation,
+/// including the MetaData, and partially, Enumerable extensions.
+abstract contract ERC721 {
+    /*///////////////////////////////////////////////////////////////
+                            EVENTS
+    //////////////////////////////////////////////////////////////*/
+
+    event Transfer(address indexed from, address indexed to, uint256 indexed tokenId);
+
+    event Approval(address indexed owner, address indexed spender, uint256 indexed tokenId);
+
+    event ApprovalForAll(address indexed owner, address indexed operator, bool approved);
+
+    /*///////////////////////////////////////////////////////////////
+                            METADATA STORAGE
+    //////////////////////////////////////////////////////////////*/
+    
+    string public name;
+
+    string public symbol;
+
+    /*///////////////////////////////////////////////////////////////
+                            ERC-721 STORAGE
+    //////////////////////////////////////////////////////////////*/
+    
+    uint256 public totalSupply;
+    
+    mapping(address => uint256) public balanceOf;
+
+    mapping(uint256 => address) public ownerOf;
+
+    mapping(uint256 => string) public tokenURI;
```

Comment:
Hello, any chance we can get rid of this? This is not very gas-optimal, and it's also not generic enough to be used by actual contracts. @abcoathup made a comment [here](https://github.com/Rari-Capital/solmate/pull/6#pullrequestreview-752239734) with a suggestion, I'm happy to add the change if you have no time to do it.

Reply:
agreed 👍

Reply:
taking a look tonight as I'm using the contract and gonna have to build it in anyways, will share